### PR TITLE
Fix newlines being eaten by rich text widget

### PR DESCRIPTION
**Pull request by millerdev**
http://manage.dimagi.com/default.asp?184602

Best reviewed by commit.

@orangejenny 
cc @emord (you should review this when you get back from vacation)

### CODE REVIEW

**Comment by millerdev** [comment]
cc @kaapstorm

**Comment by kaapstorm** [comment]
:+1: Awesome

File: src/expressionEditor.js [hunk 1]
```diff
@@ -79,18 +79,22 @@ define([
...
+                } else {
+                    input.val(val);
+                }
+            },
+            editor;
 
         var getExpression = function(input) {
             if (options.mug.supportsRichText()) {
-                return richText.fromRichText(input.ckeditor().editor.getData());
+                return richText.fromRichText(editor.getData());
```
Comment (orangejenny) [inline src/expressionEditor.js]: This line throws a js error when you go to the advanced expression editor.

Comment (orangejenny) [inline src/expressionEditor.js]: (editor is undefined)

Comment (millerdev) [inline src/expressionEditor.js]: Thanks for finding this. I'll look into it.

File: src/richText.js [hunk 2]
```diff
@@ -213,55 +216,130 @@ define([
...
-     * Takes a value from xml and replaces xpaths with bubbles to display in the
-     * editor
+     * Replace <output> tags with bubble markup
      *
-     * This does not add any <p> or <br> tags, only newlines as used in xml
+     * @param withClose - Create bubbles with close buttons if true.
+     * @param escape - If true, escape HTML except for bubble markup.
+     */
+    function bubbleOutputs(text, form, withClose, escape) {
+        var el = $('<div>').html(text),
```
Comment (emord) [inline src/richText.js]: text should be escaped like https://github.com/dimagi/Vellum/pull/554/files#diff-85ed993f6f09d78b2af63dbea82386a3R333

selection_001

**Comment by millerdev** [comment]
@emord @orangejenny updated.

Majorly refactored so CKEditor is now encapsulated in an editor object that provides a more convenient interface for vellum. This would theoretically allow us to more easily swap out CKEditor with another editor in the future if that becomes necessary.

I think this should now be easy enough to review with the PR diff rather than by commit, although the commit messages may have some useful context.

File: src/widgets.js [hunk 3]
```diff
@@ -248,112 +240,37 @@ define([
...
+        mug.on('teardown-mug-properties', function() {
+            editor.destroy();
+        }, null, "teardown-mug-properties");
 
-            editor.on('dataReady', function (e) {
-                addCloseButton(widget, widget.input);
-                addPopovers(widget.input);
-            });
+        editor.on('change', function() {
+            widget.handleChange();
```
Comment (emord) [inline src/widgets.js]: I think these could be:
```
mug.on('teardown-mug-properties', editor.destroy, null, 'teardown-mug-properties');
editor.on('change', widget.handleChange);
```

File: tests/utils.js [hunk 4]
```diff
@@ -344,8 +352,42 @@ define([
...
+     * Usage:
+     *
+     *  it("will always fail", function (done) {
+     *      someAsyncFunction(asyncatch(function (arg) {
+     *          doSomethingThatThrowsAnError();
+     *          done();
+     *      }));
+     *  });
+     */
+    function asyncatch(fn) {
```
Comment (emord) [inline tests/utils.js]: `asyncCatch` ? combining the c's looks a little strange to me

**Comment by emord** [comment]
Just non blocking comments left. Looks good to merge pending your decision on `CKEDITOR_BASEPATH`

File: src/richText.js [hunk 5]
```diff
@@ -22,51 +22,203 @@
  *
  * Any other HTML has undefined behavior
  */
+
+(function () {
+    // set CKEditor base path before loading ckeditor
+    var require = window.requirejs,
+        path = require.toUrl("vellum/../lib/ckeditor/").replace(/\?.*/, "");
+    window.CKEDITOR_BASEPATH = path;
```
Comment (millerdev) [inline src/richText.js]: Tested locally in standalone mode (built and not) and in local HQ with `VELLUM_DEBUG` set to each of `dev`, `dev-min`, and not set.

File: tests/utils.js (hunk 4 above)
Comment (millerdev) [inline tests/utils.js]: I vacillated on that one. Going to leave as is.

File: src/mugs.js [hunk 6]
```diff
@@ -356,7 +356,8 @@ define([
                 .replace("$2", property)
                 .replace("$3", String(vis)));
         },
-        getDisplayName: function (lang) {
+        getDisplayName: function (lang, escape) {
+            if (escape === undefined) { escape = true; }
```
Comment (emord) [inline src/mugs.js]: I don't think you overwrite `undefined` anymore. Do you think we should stop using `_.isUndefined` in future code?

Comment (millerdev) [inline src/mugs.js]: yes. I read up on it and `x === undefined` should be safe. I've stopped using `_.isUndefined` because I think it's ugly.

**Pull request merged**